src/protocol/sort/apply_sort/shuffle_objects.rs

```diff
@@ -0,0 +1,373 @@
+/// ii)  2 helpers apply the permutation to their shares
+/// iii) reshare to `to_helper`
+async fn shuffle_or_unshuffle_once<F, S, C, I>(
+    mut input: Vec<I>,
+    random_permutations: (&[u32], &[u32]),
+    shuffle_or_unshuffle: ShuffleOrUnshuffle,
+    ctx: &C,
+    which_step: ShuffleStep,
+) -> Result<Vec<I>, Error>
+where
+    C: Context<F, Share = S> + Send,
+    F: Field,
+    I: Resharable<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let to_helper = shuffle_for_helper(which_step);
+    let ctx = ctx.narrow(&which_step);
+
+    if to_helper != ctx.role() {
+        let permutation_to_apply = if to_helper.peer(Direction::Left) == ctx.role() {
+            random_permutations.0
+        } else {
+            random_permutations.1
+        };
+
+        match shuffle_or_unshuffle {
+            ShuffleOrUnshuffle::Shuffle => apply_inv(permutation_to_apply, &mut input),
+            ShuffleOrUnshuffle::Unshuffle => apply(permutation_to_apply, &mut input),
+        }
+    }
+    reshare_objects(&input, ctx, to_helper).await
+}
+
+#[embed_doc_image("shuffle", "images/sort/shuffle.png")]
+/// Shuffle calls `shuffle_or_unshuffle_once` three times with 2 helpers shuffling the shares each time.
+/// Order of calling `shuffle_or_unshuffle_once` is shuffle with (H2, H3), (H3, H1) and (H1, H2).
+/// Each shuffle requires communication between helpers to perform reshare.
+/// Infrastructure has a pre-requisite to distinguish each communication step uniquely.
+/// For this, we have three shuffle steps one per `shuffle_or_unshuffle_once` i.e. Step1, Step2 and Step3.
+/// The Shuffle object receives a step function and appends a `ShuffleStep` to form a concrete step
+/// ![Shuffle steps][shuffle]
+pub async fn shuffle_object_shares<C, F, I, S>(
+    input: Vec<I>,
+    random_permutations: (&[u32], &[u32]),
+    ctx: C,
+) -> Result<Vec<I>, Error>
+where
+    C: Context<F, Share = S> + Send,
+    F: Field,
+    I: Resharable<F, Share = S>,
+    S: SecretSharing<F>,
+{
+    let input = shuffle_or_unshuffle_once(
+        input,
+        random_permutations,
+        ShuffleOrUnshuffle::Shuffle,
+        &ctx,
+        Step1,
+    )
+    .await?;
+    let input = shuffle_or_unshuffle_once(
+        input,
+        random_permutations,
+        ShuffleOrUnshuffle::Shuffle,
+        &ctx,
+        Step2,
+    )
+    .await?;
+    shuffle_or_unshuffle_once(
+        input,
+        random_permutations,
+        ShuffleOrUnshuffle::Shuffle,
+        &ctx,
+        Step3,
+    )
+    .await
+}
+
+#[embed_doc_image("unshuffle", "images/sort/unshuffle.png")]
+/// Unshuffle calls `shuffle_or_unshuffle_once` three times with 2 helpers shuffling the shares each time in the opposite order to shuffle.
+/// Order of calling `shuffle_or_unshuffle_once` is shuffle with (H1, H2), (H3, H1) and (H2, H3)
+/// ![Unshuffle steps][unshuffle]
+pub async fn unshuffle_object_shares<F, S, C, I>(
```

Comment:
It might be enough to call these functions "shuffle" and "unshuffle".  Or, "shuffle_shares" and "unshuffle_shares".  The object part is implied.

Reply:
Removed unshuffle since we dont need it for applying sort